agent/xds/delta_test.go

```diff
@@ -45,132 +37,228 @@ func TestServer_DeltaAggregatedResources_v3_BasicProtocol_TCP(t *testing.T) {
-		TypeUrl: ListenerType,
-		Nonce:   hexString(3),
-		Resources: makeTestResources(t,
-			makeTestListener(t, snap, "tcp:public_listener"),
-			makeTestListener(t, snap, "tcp:db"),
-			makeTestListener(t, snap, "tcp:geo-cache"),
-		),
-	})
+		// It also (in parallel) issues the endpoint ACK
+		envoy.SendDeltaReqACK(t, EndpointType, 2, true, nil)
 
-	// cleanup unused resources now that we've created/updated relevant things
-	assertDeltaResponseSent(t, envoy.deltaStream.sendCh, &envoy_discovery_v3.DeltaDiscoveryResponse{
-		TypeUrl: EndpointType,
-		Nonce:   hexString(4),
-		RemovedResources: []string{
-			"fake-endpoints", // correcting the errant subscription
-		},
+		// And should get a response immediately.
+		assertDeltaResponseSent(t, envoy.deltaStream.sendCh, &envoy_discovery_v3.DeltaDiscoveryResponse{
+			TypeUrl: ListenerType,
+			Nonce:   hexString(3),
+			Resources: makeTestResources(t,
+				makeTestListener(t, snap, "tcp:public_listener"),
+				makeTestListener(t, snap, "tcp:db"),
+				makeTestListener(t, snap, "tcp:geo-cache"),
+			),
+		})
+
+		// cleanup unused resources now that we've created/updated relevant things
+		assertDeltaResponseSent(t, envoy.deltaStream.sendCh, &envoy_discovery_v3.DeltaDiscoveryResponse{
+			TypeUrl: EndpointType,
+			Nonce:   hexString(4),
+			RemovedResources: []string{
+				"fake-endpoints", // correcting the errant subscription
+			},
+		})
+
+		// And no other response yet
+		assertDeltaChanBlocked(t, envoy.deltaStream.sendCh)
+
+		// ACKs the listener
+		envoy.SendDeltaReqACK(t, ListenerType, 3, true, nil)
+
+		// ACK the endpoint removal
+		envoy.SendDeltaReqACK(t, EndpointType, 4, true, nil)
+
+		// If we re-subscribe to something even if there are no changes we get a
+		// fresh copy.
+		envoy.SendDeltaReq(t, EndpointType, &envoy_discovery_v3.DeltaDiscoveryRequest{
+			ResourceNamesSubscribe: []string{
+				"geo-cache.default.dc1.query.11111111-2222-3333-4444-555555555555.consul",
+			},
+		})
+
+		assertDeltaResponseSent(t, envoy.deltaStream.sendCh, &envoy_discovery_v3.DeltaDiscoveryResponse{
+			TypeUrl: EndpointType,
+			Nonce:   hexString(5),
+			Resources: makeTestResources(t,
+				makeTestEndpoints(t, snap, "tcp:geo-cache"),
+			),
+		})
+
+		envoy.SendDeltaReqACK(t, EndpointType, 5, true, nil)
+
+		// And no other response yet
+		assertDeltaChanBlocked(t, envoy.deltaStream.sendCh)
 	})
 
-	// And no other response yet
-	assertDeltaChanBlocked(t, envoy.deltaStream.sendCh)
+	deleteAllButOneEndpoint := func(snap *proxycfg.ConfigSnapshot, svc, targetID string) {
+		snap.ConnectProxy.ConfigSnapshotUpstreams.WatchedUpstreamEndpoints[svc][targetID] =
+			snap.ConnectProxy.ConfigSnapshotUpstreams.WatchedUpstreamEndpoints[svc][targetID][0:1]
+	}
 
-	// ACKs the listener
-	envoy.SendDeltaReqACK(t, ListenerType, 3, true, nil)
+	runStep(t, "envoy randomly decides to unsubscribe from an endpoint", func(t *testing.T) {
```

Comment:
I think renaming the subtest to include the expected result could make the test case clearer:
```suggestion
	runStep(t, "avoid sending config for unsubscribed resource", func(t *testing.T) {
```